src/SArray.jl

```diff
@@ -50,6 +50,34 @@ end
     end
 end
 
+@generated function (::Type{SArray{S, T}})(gen::Base.Generator) where {S <: Tuple, T}
+    stmts = [:(Base.@_inline_meta)]
+    args = []
+    iter = :(iterate(gen))
+    for i in CartesianIndices(size_to_tuple(S))
+        el = Symbol(:el, i)
+        push!(stmts, :(($el,st) = $iter))
+        push!(args, el)
+        iter = :(iterate(gen,st))
+    end
+    push!(stmts, :(SArray{S, T}($(args...))))
```

Comment:
I think we need something like `$iter === nothing || error(...)` here.

Reply:
Can do. (I left it out deliberately.)